cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -819,12 +819,30 @@ Type getTypeImpl(Type t, int indirectionIndex) {
     // We need to avoid the case where `stripPointer(t) = t` (which can happen on
     // iterators that specify a `value_type` that is the iterator itself). Such a type
     // would create an infinite loop otherwise. For these cases we simply don't produce
-    // a result for `getType`.
+    // a result for `getTypeImpl`.
     stripped.getUnspecifiedType() != t.getUnspecifiedType() and
-    result = getTypeImpl(stripped, indirectionIndex - 1)
+    result = getTypeImpl0(stripped, indirectionIndex - 1)
   )
 }
 
+/**
+ * INTERNAL: Do not use.
+ *
+ * Returns `t`, but stripped of the outer-most `indirectionIndex` number of indirections.
+ *
+ * If `indirectionIndex` cannot be stripped off `t`, an `UnknownType` is returned.
+ */
+bindingset[indirectionIndex]
+Type getTypeImpl(Type t, int indirectionIndex) {
+  result = getTypeImpl0(t, indirectionIndex)
+  or
+  // If we cannot produce the right type we return an error type.
+  // This can sometimes happen when we don't know the real
+  // type of a void pointer.
+  not exists(getTypeImpl0(t, indirectionIndex)) and
+  result instanceof UnknownType
+}
```

Comment:
I'm a bit worried about the performance of this, even with `bindingset[indirectionIndex]`, e.g. if `indirectionIndex` can take a known range of values and `t` is not bound.  You may well know better than I do here.  In any case, DCA should let us know if there's a serious problem...

Reply:
I think we should be good here. Most things should be covered by that first disjunct, so there won't be too many things hitting that second disjunct anyway. Maybe I should've actually put a `bindingset` annotation on `t` as well though (as it will actually be bound by all callers anyway).

Reply:
Looking at it today, I think I agree it's unlikely to be an issue.  Thanks!